### PR TITLE
add design doc for rescheduling

### PR DESCRIPTION
**Pull request by Thor-wl**
ref: https://github.com/volcano-sh/volcano/issues/1777
Signed-off-by: Thor-wl <13164644535@163.com>

### CODE REVIEW

File: docs/design/descheduler.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    `Label Filter` will filter jobs with specified labels. Then the rescheduling process only works on the result set.
+
+### Metrics
+All the decisions made by `DeScheduler` will consider the metrics from `Prometheus`. Namely, `DeScheduler` will list the
+real node resource utilization and pod distribution instead of request resource. Basically, usage of`CPU` and `Memroy` 
+will be collected. Other resource such as `GPU` can be extended.
+
+## Implementation Plan
+1. Fork from [kubernetes-sigs/descheduler](https://github.com/kubernetes-sigs/descheduler) to implement the basic ability.
```
Comment (k82cn) [inline docs/design/descheduler.md:68]: Please do not fork it; it's hard to maintain another repo :(

**Comment by hwdef** [comment]
If the resource usage rate of a pod changes frequently, will this pod be killed frequently?

**Comment by hwdef** [comment]
Can this feature coexist with gang normally? For example, what are the consequences of deleting a normally running pod?
If a pod of MPI is deleted, the entire job will restart.

**Comment by hwdef** [comment]
https://github.com/petuum/adaptdl
https://www.usenix.org/conference/osdi21/presentation/qiao

A digression, if we can implement such a scheduling method, it would be best.

File: docs/design/descheduler.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+## Motivation
+As what [Issue1777](https://github.com/volcano-sh/volcano/issues/1777) describes, **Rescheduling** is important for the 
+following reasons:
+* Unbalanced resource utilization due to unreasonable scheduling strategies and dynamic changes in jobs' lifecycle.
+* Node status changes such as add/remove nodes, pod/node taint/affinity changes.
+
+In order to rebalance the cluster resource utilization among nodes, introduce **DeScheduler** to achieve these goals:
+* Rescheduling pods based on real resource utilization instead of request resource.
+* Support custom configured rescheduling strategies.
```
Comment (william-wang) [inline docs/design/descheduler.md:13]: Add the inscope and out of scope to let users know what will be done in this project.

File: docs/design/descheduler.md (hunk 1 above)
Comment (Thor-wl) [reply docs/design/descheduler.md:68]: I'm seeking a way to implement custom strategies without forking the upstream repository. Waiting for the answer from https://github.com/kubernetes-sigs/descheduler/issues/686

**Comment by Thor-wl** [comment]
> 

Will get the average rate for the latest some time, for example, 1min

**Comment by Thor-wl** [comment]
> If the resource usage rate of a pod changes frequently, will this pod be killed frequently?

Will get the average rate for the latest some time, for example, 1min

**Comment by Thor-wl** [comment]
> Can this feature coexist with gang normally? For example, what are the consequences of deleting a normally running pod? If a pod of MPI is deleted, the entire job will restart.

Will implemente the group rescheduling strategy later. The doc has mentioned these points.

File: docs/design/descheduler.md (hunk 2 above)
Comment (Thor-wl) [reply docs/design/descheduler.md:13]: Roadmap added to the `Implementation Plan` chapter

File: docs/design/descheduler.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+          offlineOnly: true
+          lowPriorityFirst: true
+  - plugins:
+      - name: overcommit
+      - name: drf
+      - name: predicates
+      - name: proportion
+      - name: nodeorder
+      - name: binpack
+rescheduling:    ## configurations for rescheduling
```
Comment (k82cn) [inline docs/design/descheduler.md:84]: it should be a plugin, instead of importing all code from descheduler.

Comment (Thor-wl) [reply docs/design/descheduler.md]: Not import all code from descheduler, reimplement part of its functions.

File: docs/design/rescheduling.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          offlineOnly: true
+          lowPriorityFirst: true
+  - plugins:
+      - name: overcommit
+      - name: drf
+      - name: predicates
+      - name: proportion
+      - name: nodeorder
+      - name: binpack
+rescheduling:
```
Comment (k82cn) [inline docs/design/rescheduling.md:84]: Are we going to implement this feature by plugin?

Comment (Thor-wl) [reply docs/design/rescheduling.md:84]: > 

Yes. one action named `shuffle` works to do rescheduling and one plugin named `strategy` to provide the rescheduling algorithms.

File: docs/design/rescheduling.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+### Metrics
+All the decisions made by rescheduling strategies will consider the metrics from `Prometheus`. Namely, Volcano will 
+list the real node resource utilization and pod distribution instead of request resource. Basically, usage of`CPU` and 
+`Memroy` will be collected. Other resource such as `GPU` can be extended.
+
+## Implementation(To Be Discussed)
+```yaml
+## Configuration Option 1
+actions: "enqueue, allocate, backfill, shuffle"  ## add 'shuffle' at the end of the actions
+action_configurations:       ## configurations for actions
```
Comment (k82cn) [inline docs/design/rescheduling.md:68]: why do we have such configuration for action？

File: docs/design/rescheduling.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      - name: gang
+      - name: conformance
+  - plugins:
+      - name: overcommit
+      - name: drf
+      - name: predicates
+      - name: proportion
+      - name: nodeorder
+      - name: binpack  
+rescheduling:            ## configurations for rescheduling
```
Comment (k82cn) [inline docs/design/rescheduling.md:110]: what is `rescheduling`, action or plugin?

File: docs/design/rescheduling.md (hunk 5 above)
Comment (k82cn) [reply docs/design/rescheduling.md:68]: let's move this configuration to plugin

File: docs/design/rescheduling.md (hunk 6 above)
Comment (Thor-wl) [reply docs/design/rescheduling.md:110]: > what is `rescheduling`, action or plugin?

Neithor is action or plugins. In fact, it works for 'shuffle' action. Unfortunately, there is no configuration options for actions. So I have to add these configurations.

File: docs/design/rescheduling.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      - name: predicates
+      - name: proportion
+      - name: nodeorder
+      - name: binpack
+```
+
+* Load and parse user configurations about rescheduling.
+* Update the cache by metrics collected by `Prometheus`.
+* Create a new action named "shuffle", implement the workflow above.
+* Create a new plugin named "rescheduling", implement the strategies above.
```
Comment (william-wang) [inline docs/design/rescheduling.md:97]: What's your plan for dynamic scheduling and rescheduling. combine them in one plugin or two plugin?

File: docs/design/rescheduling.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    `MoreReplicasFirst`, abbreviated as `MRF`, will sort workloads by replicas number. Workloads with the most replicas
+will be rescheduled first. This strategy is friendly to `gang scheduling` for it will consider the `minAvailable` in 
+volcano jobs.
+
+### Metrics
+All the decisions made by rescheduling strategies will consider the metrics from `Prometheus`. Namely, Volcano will 
+list the real node resource utilization and pod distribution instead of request resource. Basically, usage of`CPU` and 
+`Memroy` will be collected. Other resource such as `GPU` can be extended.
+
+## Implementation(Beta)
```
Comment (william-wang) [inline docs/design/rescheduling.md:64]: Would you declare explictly which parts/configurations/stratages will be supported in next release(1.6) and which parts are long term feature? Users might be confused about the Volcano ablility while they are reading the doc.

File: docs/design/rescheduling.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+## Configuration Option 
+actions: "enqueue, allocate, backfill, shuffle"  ## add 'shuffle' at the end of the actions
+tiers:
+  - plugins:
+      - name: priority
+      - name: gang
+      - name: conformance
+      - name: rescheduling       ## rescheduling plugin
+        arguments:
+          interval: 5m
```
Comment (william-wang) [inline docs/design/rescheduling.md:75]: It's better to add instruction about the format of `interval`. And the interval should be larger than scheduling period at least.

**Review by william-wang — CHANGES_REQUESTED** [review]

File: docs/design/rescheduling.md (hunk 7 above)
Comment (Thor-wl) [reply docs/design/rescheduling.md:97]: Will implement them as two plugins. Besides, the logic of part of functions for them is absolutely different.

File: docs/design/rescheduling.md (hunk 8 above)
Comment (Thor-wl) [reply docs/design/rescheduling.md:64]: OK. Let me add the plan details.

File: docs/design/rescheduling.md (hunk 9 above)
Comment (Thor-wl) [reply docs/design/rescheduling.md:75]: That's reasonable. Let me add the comment.

File: docs/design/rescheduling.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+## Configuration Option 
+actions: "enqueue, allocate, backfill, shuffle"  ## add 'shuffle' at the end of the actions
+tiers:
+  - plugins:
+      - name: priority
+      - name: gang
+      - name: conformance
+      - name: rescheduling       ## rescheduling plugin
+        arguments:
+          interval: 5m           ## optional, the strategies will be called in this duration periodcally. 5 minuters by default.
```
Comment (william-wang) [inline docs/design/rescheduling.md:79]: How to configure threshhold for cpu, mem?

Comment (Thor-wl) [reply docs/design/rescheduling.md:79]: I've submitted another commit to give an configuration example for lowNodeUtilization strategy.

**Review by william-wang — APPROVED** [review]
/lgtm

**Comment by jiangxiaobin96** [comment]
Is there any possibility that some pods are evicted frequently due to always satisfy the rescheduling plugin requirements? I mean if the interval time should be set bigger.

**Pull request merged**